src/vec.rs

```diff
@@ -0,0 +1,86 @@
+use rustc::lint::*;
+use rustc::middle::ty::TypeVariants;
+use rustc_front::hir::*;
+use utils::{BOX_NEW_PATH, VEC_FROM_ELEM_PATH};
+use utils::{is_expn_of, match_path, snippet, snippet_opt, span_lint_and_then};
+
+/// **What it does:** This lint warns about using `&vec![..]` when using `&[..]` would be possible.
+/// It is `Warn` by default.
+///
+/// **Why is this bad?** This is less efficient.
+///
+/// **Known problems:** None.
+///
+/// **Example:** 
+/// ```rust, ignore
+/// foo(&vec![1, 2])
+/// ```
+declare_lint! {
+    pub USELESS_VEC,
+    Warn,
+    "useless `vec!`"
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct UselessVec;
+
+impl LintPass for UselessVec {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(USELESS_VEC)
+    }
+}
+
+impl LateLintPass for UselessVec {
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        // search for `&!vec[_]` expressions where the adjusted type is `&[_]`
+        if_let_chain!{[
+            let TypeVariants::TyRef(_, ref ty) = cx.tcx.expr_ty_adjusted(expr).sty,
+            let TypeVariants::TySlice(..) = ty.ty.sty,
+            let ExprAddrOf(_, ref addressee) = expr.node,
+            let ExprCall(ref fun, ref args) = addressee.node,
+            let ExprPath(_, ref path) = fun.node,
+            let Some(span) = is_expn_of(cx, fun.span, "vec")
+        ], {
+            if match_path(path, &VEC_FROM_ELEM_PATH) && args.len() == 2 {
```

Comment:
Could we restructure this into a util that takes an expression and provides you with the list of elements of the vector pre-expansion? Would be useful for other lints.